feat(fs): support FullscreenOptions

apmorton opened the pull request:
## Description
Starting in chrome 71, the Fullscreen API is being shipped unprefixed.
The behavior on android (and possibly other platforms I have not personally checked) has also changed slightly.

It may be hard to see on a white background, but you should see the "soft keys" stay visible on the right hand side of this screenshot when in fullscreen.  This is observed on a Pixel 2 XL on the latest security patch level with the latest chrome version.
![image](https://user-images.githubusercontent.com/63636/54222819-85c7ed00-44cc-11e9-8540-1b6d567a73fd.png)

Starting with Chrome 71, `webkitRequestFullscreen()` is equivalent to `requestFullscreen({ navigationUI: 'hide' })` [as can be seen in the source](https://chromium-review.googlesource.com/c/chromium/src/+/1204595/4/third_party/blink/renderer/core/fullscreen/element_fullscreen.cc)

I have added an option to the player called `navigationUI` and defaulted it to `hide`, such that simply updating video.js on "broken" sites should resolve the issue, but still allow for the "new" behavior introduced by chrome 71 if desired.

## Requirements Checklist
- [x] Feature implemented / Bug fixed
- [ ] If necessary, more likely in a feature request than a bug fix
  - [x] Change has been verified in an actual browser (Chome, Firefox, IE)
  - [ ] Unit Tests updated or fixed
  - [x] Docs/guides updated
  - [ ] Example created ([starter template on JSBin](https://codepen.io/gkatsev/pen/GwZegv?editors=1000#0))
- [ ] Reviewed by Two Core Contributors

## References
https://groups.google.com/a/chromium.org/forum/#!msg/blink-dev/ODzbWn-xRrQ/dhkJzD8-CgAJ
https://www.chromestatus.com/feature/6596356319739904
https://www.chromestatus.com/feature/5188650908254208
https://developer.mozilla.org/en-US/docs/Web/API/FullscreenOptions
https://github.com/whatwg/fullscreen/pull/129

gkatsev commented:
Thanks for the PR, I'm a bit worried with flipping the `prefixedAPI` default as we have very little bandwidth currently to re-run the entire testing we did around it. Our current fs support is a bit flaky and we need to refactor it.

I also am not sure that navigationUI should be a player option. ~Likely, we can follow suit, and just add it as an option in the method.~ As I was typing that, I realized that we have a few other places where we call requestFullscreen on the behalf of the user. I think rather than `navigationUI`, we have an option called `fullscreenOptions` that we then pass where appropriate, this also makes it that if new options are added, we won't need to update the player. And we probably still want to let the user be able to pass in `fullscreenOptions` to `requestFullscreen` which will take precedence over the player options.

apmorton commented:
I didn't explicitly call it out in the PR, but in the commit message I justified the `prefixedAPI` changes.

From commit message follows:

Additionally invert misleading fullscreen-api boolean `prefixedAPI` which was true when *not* prefixed 
and false when prefixed.
This may have been an unintentional change in 1c0fa32.


As far as I could tell from grepping the entire repo it was only used in a single place (before I added another usage).  I flipped *everything*, not just the default value.  It should be functionally identically to before.  The change is purely to make the name match the value.

In the existing code, `prefixedAPI === true` when `requestFullscreen` is used and `prefixedAPI === false` when `webkitRequestFullscreen` is used.  I found this misleading and corrected it.

If you still take issue with that admittedly unrelated change, I have no problem removing it.

Will prototype the other changes and update the PR.

apmorton commented:
@gkatsev any comments on the updated implementation?

gkatsev commented:
I'll probably have time to take a look early next week. Thanks for including tests!

gkatsev commented:
The changes look good, though, we should keep fsApi on the player private. Another thing to consider is whether and how this plays with https://github.com/videojs/video.js/pull/5757 and https://github.com/videojs/video.js/issues/5885.

apmorton commented on src/js/player.js:
```diff
@@ -354,6 +354,9 @@ class Player extends Component {
     this.log = createLogger(this.id_);
 
+    // Hold our own reference to fullscreen api so it can be mocked in tests
+    this.fsApi = FullscreenApi;
```
guessing `_` indicates the member is "private"?

The `fsApi` member is accessed from the `FullscreenToggle` class - do you still want this to be suffixed with `_`?

I added a line note where it is accessed

gkatsev replied:
Yeah, private, though to the library rather than to the class/object; it's fine if the FullscreenToggle uses it. We just don't want to officially expose it publicly as part of this PR. The suffix isn't perfect, but it's a good enough convention.

apmorton commented on src/js/control-bar/fullscreen-toggle.js:
```diff
@@ -26,7 +25,7 @@ class FullscreenToggle extends Button {
     this.on(player, 'fullscreenchange', this.handleFullscreenChange);
 
-    if (document[FullscreenApi.fullscreenEnabled] === false) {
+    if (document[player.fsApi.fullscreenEnabled] === false) {
```
`fsApi` accessed outside player here

apmorton commented:
As for #5757, it will conflict, but only simple fixes need to be made to whichever is merged last.

You probably noticed, but there are already some conflicts with master and this PR - but they are trivial fixes.

I will hold off on rebasing this PR till you advise on the `_` question.

apmorton commented:
rebased on master and `fsApi` changed to `fsApi_`

apmorton commented:
@gkatsev ping

this PR again has conflicts with master, but they don't look too hard to resolve.

If you don't have any other issues with the implementation I can rebase so it can be merged.

gkatsev commented:
Hey, it looks good. A rebase would be welcome.
Still thinking about how to reconcile this PR along with #5757 and #5885 but it won't affect this PR, I don't think.

apmorton commented:
@gkatsev done

gkatsev commented:
@apmorton thanks!

apmorton commented:
@gkatsev ping

can somebody *please* merge this?

gkatsev commented:
I appreciate your extreme patience with us. We'll be merging it this week and we'll have a release shortly after.

gkatsev commented:
Thanks @apmorton! Sorry it's taken so long. FYI, we've decided to change the option format slightly to better support a fullscreen element and a fullscreen handler option. See #6054.
